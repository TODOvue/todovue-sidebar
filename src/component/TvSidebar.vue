<template>
  <div class="tv-sidebar-body">
    <div class="tv-sidebar">
      <template v-if="isImage">
        <template v-if="clickable">
          <component :is="linkComponent" :to="data.image.link">
            <img
              :src="data.image.src"
              :alt="data.image.alt"
              style="cursor: pointer"
            />
          </component>
        </template>
        <template v-else>
          <img :src="data.image.src" :alt="data.image.alt" />
        </template>
      </template>
      <template v-else-if="isLabel">
        <div class="tv-sidebar-title">
          <h1>{{ data.title }}</h1>
          <div class="tv-sidebar-title-separator"></div>
        </div>
        <div class="tv-sidebar-content-label">
          <tv-label
            v-for="label in limitedList('labels')"
            :key="label.id"
            :color="label.color"
            @click="clickLabel(label)"
          >
            {{ label.name }}
          </tv-label>
        </div>
      </template>
      <template v-else>
        <div class="tv-sidebar-title">
          <h1>{{ data.title }}</h1>
          <div class="tv-sidebar-title-separator"></div>
        </div>
        <div class="tv-sidebar-content">
          <ol
            class="tv-sidebar-content-ol"
            v-for="(item, index) in limitedList('list')"
            :key="item.id"
          >
            <li class="tv-sidebar-content-li">
              <component :is="linkComponent" :to="item.link" class="pointer">
                <span>{{ index + 1 }}.</span>{{ item.title }}
              </component>
            </li>
          </ol>
        </div>
      </template>
    </div>
  </div>
</template>

<script>
import TvLabel from "todovue-label";

export default {
  name: "TvSidebar",
  props: {
    data: {
      type: Object,
      default: () => ({}),
    },
    linkTag: {
      type: String,
      default: "router-link",
    },
    isImage: {
      type: Boolean,
      default: false,
    },
    isLabel: {
      type: Boolean,
      default: false,
    },
    limit: {
      type: Number,
      default: 0,
    },
    clickable: {
      type: Boolean,
      default: false,
    },
  },
  components: {
    TvLabel,
  },
  computed: {
    linkComponent() {
      return this.linkTag === "nuxt-link" ? "nuxt-link" : "router-link";
    },
  },
  methods: {
    clickLabel(label) {
      this.$emit("clickLabel", label);
    },
    limitedList(property) {
      if (!this.data[property]) return [];
      let limit = this.limit > 0 ? this.limit : this.data[property].length;
      return this.data[property].slice(0, limit);
    },
  },
};
</script>

<style></style>
