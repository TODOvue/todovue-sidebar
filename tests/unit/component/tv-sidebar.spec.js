import { shallowMount } from "@vue/test-utils";
import TvSidebar from "@/component/TvSidebar.vue";
import TvLabel from "todovue-label";

describe("TvButton Component", () => {
  it("should matches snapshot", () => {
    const wrapper = shallowMount(TvSidebar);
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("Render the image correctly when 'isImage' is true.", () => {
    const data = {
      image: {
        src: "https://ejemplo.com/imagen.png",
        alt: "Ejemplo de imagen",
        link: "https://ejemplo.com/pagina",
      },
    };
    const wrapper = shallowMount(TvSidebar, {
      propsData: {
        data,
        isImage: true,
      },
    });
    const imagen = wrapper.find("img");
    expect(imagen.exists()).toBe(true);
    expect(imagen.attributes("src")).toBe(data.image.src);
    expect(imagen.attributes("alt")).toBe(data.image.alt);
  });

  test("The 'data' property has the default value of an empty object.", () => {
    const wrapper = shallowMount(TvSidebar);
    expect(wrapper.vm.data).toEqual({});
  });

  test("The 'linkTag' property has the default value of \"router-link\".", () => {
    const wrapper = shallowMount(TvSidebar);
    expect(wrapper.vm.linkTag).toBe("router-link");
  });

  test("The 'isImage' property has the default value of 'false'.", () => {
    const wrapper = shallowMount(TvSidebar);
    expect(wrapper.vm.isImage).toBe(false);
  });

  test("The TvLabel component is being imported correctly.", () => {
    const wrapper = shallowMount(TvSidebar);
    expect(wrapper.findComponent(TvLabel)).toBeTruthy();
  });

  test("The 'limitedList' method returns an empty array when the corresponding property in 'data' is null or undefined.", () => {
    const wrapper = shallowMount(TvSidebar);
    expect(wrapper.vm.limitedList("labels")).toEqual([]);
  });

  test("The 'limitedList' method returns an array with limited elements when the 'limit' property is greater than zero.", () => {
    const data = {
      list: [
        { id: 1, title: "Item 1", link: "/item/1" },
        { id: 2, title: "Item 2", link: "/item/2" },
        { id: 3, title: "Item 3", link: "/item/3" },
      ],
    };
    const limit = 2;
    const wrapper = shallowMount(TvSidebar, {
      propsData: {
        data,
        limit,
      },
    });
    expect(wrapper.vm.limitedList("list")).toHaveLength(limit);
  });

  test("The 'clickLabel' event is emitted correctly when a label is clicked.", () => {
    const label = { id: 1, name: "Etiqueta", color: "#f00" };
    const wrapper = shallowMount(TvSidebar);
    wrapper.vm.clickLabel(label);
    expect(wrapper.emitted().clickLabel).toBeTruthy();
    expect(wrapper.emitted().clickLabel[0][0]).toBe(label);
  });
});
