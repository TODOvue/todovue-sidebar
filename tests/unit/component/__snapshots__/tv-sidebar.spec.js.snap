// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TvButton Component should matches snapshot 1`] = `
<div class="tv-sidebar-body">
  <div class="tv-sidebar">
    <div class="tv-sidebar-title">
      <h1></h1>
      <div class="tv-sidebar-title-separator"></div>
    </div>
    <div class="tv-sidebar-content"></div>
  </div>
</div>
`;
